theano/tensor/nnet/tests/test_ctc.py

```diff
@@ -0,0 +1,111 @@
+
+        utt.assert_allclose(expected_grads / cost.shape[0], grad)
+        utt.assert_allclose(expected_costs, cost)
+
+    # Test obtained from Torch tutorial at:
+    # https://github.com/baidu-research/warp-ctc/blob/master/torch_binding/TUTORIAL.md
+    def test_torch_case(self):
+        # Layout, from slowest to fastest changing dimension, is (time, batchSize, inputLayerSize)
+        activations = np.asarray([[[0, 0, 0, 0, 0], [1, 2, 3, 4, 5], [-5, -4, -3, -2, -1]],
+                                  [[0, 0, 0, 0, 0], [6, 7, 8, 9, 10], [-10, -9, -8, -7, -6]],
+                                  [[0, 0, 0, 0, 0], [11, 12, 13, 14, 15], [-15, -14, -13, -12, -11]]],
+                                 dtype=np.float32)
+        # Duration of each sequence
+        activation_times = np.asarray([1, 3, 3], dtype=np.int32)
+        # Labels for each sequence
+        labels = np.asarray([[1, -1],
+                             [3, 3],
+                             [2, 3]], dtype=np.int32)
+
+        expected_costs = np.asarray([1.609437943, 7.355742931, 4.938849926],
+                                    dtype=np.float32)
+
+        grads = [[[0.2, -0.8, 0.2, 0.2, 0.2],
+                  [0.01165623125, 0.03168492019, 0.08612854034, -0.7658783197, 0.636408627],
+                  [-0.02115798369, 0.03168492019, -0.8810571432, 0.2341216654, 0.636408627]],
+                 [[0, 0, 0, 0, 0],
+                  [-0.9883437753, 0.03168492019, 0.08612854034, 0.2341216654, 0.636408627],
+                  [-0.02115798369, 0.03168492019, -0.1891518533, -0.4577836394, 0.636408627]],
+                 [[0, 0, 0, 0, 0],
+                  [0.01165623125, 0.03168492019, 0.08612854034, -0.7658783197, 0.636408627],
+                  [-0.02115798369, 0.03168492019, 0.08612854034, -0.7330639958, 0.636408627]]]
+        expected_gradients = np.asarray(grads, dtype=np.float32)
+
+        self.run_ctc(activations, labels, activation_times, expected_costs, expected_gradients)
+
+    def test_ctc(self):
+        activations = np.asarray([[[0.1, 0.6, 0.1, 0.1, 0.1], [0.1, 0.1, 0.6, 0.1, 0.1]],
+                                  [[0.6, 0.1, 0.1, 0.1, 0.1], [0.1, 0.1, 0.5, 0.2, 0.1]]],
+                                 dtype=np.float32)
+
+        activation_times = np.asarray([2, 2], dtype=np.int32)
+
+        labels = np.asarray([[1, 2], [1, 2]], dtype=np.int32)
+
+        expected_costs = np.asarray([2.962858438, 3.053659201], dtype=np.float32)
+
+        grads = [[[0.177031219, -0.7081246376, 0.177031219, 0.177031219, 0.177031219],
+                  [0.177031219, -0.8229685426, 0.291875124, 0.177031219, 0.177031219]],
+                 [[0.291875124, 0.177031219, -0.8229685426, 0.177031219, 0.177031219],
+                  [0.1786672771, 0.1786672771, -0.7334594727, 0.1974578798, 0.1786672771]]]
+
+        expected_gradients = np.asarray(grads, dtype=np.float32)
+
+        self.run_ctc(activations, labels, activation_times, expected_costs, expected_gradients)
+
```

Comment:
We also need a quick test to check that grad is not computed when it is not used (ie. optimization that removes gradient is applied correctly).

Just build a short theano function that computes cost only, and then check that `node.op.compute_grad` is false for the node with a CTC op.